# Food Balance Sheet

#### DataSet

The dataset is provided by the Food and Agriculture Organization of the United Nations and is made available in .csv format
[FoodBalanceSheets_E_Africa_NOFLAG.csv](https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv)

#### Data Information

The details in the dataset are: Area Code, Area, Item Code, Item, Element Code, Element, Unit, Y2014, Y2015, Y2016, Y2017, Y2018


In [1]:
# importing the necessary libraries
import numpy as np
import pandas as pd

#### Loading the csv file 

In [2]:
# reading it with pandas
df = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding = 'Latin-1')

In [3]:
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

# Question 1: 
### If you havethe following list
    lst =[[35,'Portugal', 94], [33, 'Argentina', 93],[ 30, 'Brazil', 92]]
    col = ['Age', 'Nationality', 'Overall']
How do you create a pandas DataFrame using the list?

In [5]:
lst =[[35,'Portugal', 94], [33, 'Argentina', 93],[ 30, 'Brazil', 92]]
col = ['Age', 'Nationality', 'Overall']

pd.DataFrame(lst, columns = col,index = [i for i in range(1,4)])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

# Question 2: 
### What type of error will this output give?
    my_tuppy = (1,2,5,8)
    my_tuppy[2] = 6

In [6]:
my_tuppy = (1,2,5,8)
my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

It gives a typeError

# Question 3:
### What is the mean and standard deviation across the whole dataset for the year2017 to 2 decimal places?

In [7]:
# The Mean
mean = round(df['Y2017'].mean(),2)
mean

140.92

In [8]:
# The Standard Deviation
std = round(df['Y2017'].std(),2)
std

1671.86

# Question 4:
### What is the total number and percentage of missing data in 2014 to 3 decimal?

In [9]:
missing_values =df['Y2014'].isnull().sum()
missing_values

1589

In [10]:
no_of_rows = df.shape[0]
no_of_rows

60943

In [12]:
percentage_missing = round((missing_values/no_of_rows*100),3)
percentage_missing

2.607

#### Total number and percentage of missing data in 2014 is 1589 and 2.607 %

# Question 5:
### Perform a groupby operation on ‘Element’. What year has the highest sum of StockVariation?

In [13]:
byElement = df.groupby('Element').sum()
byElement.head()

Area Code  Item Code  Element Code  \
Element                                                                  
Domestic supply quantity               708993   14197445      28068795   
Export Quantity                        599910   11840553      26026133   
Fat supply quantity (g/capita/day)     675050   13535000       3435732   
Feed                                   176272    3538507       7282199   
Food                                   663295   13285035      25406622   

                                         Y2014       Y2015       Y2016  \
Element                                                                  
Domestic supply quantity            1996716.35  2021493.55  2044842.70   
Export Quantity                      150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)    10225.56    10235.74    10102.77   
Feed                                 216927.89   225050.22   228958.65   
Food                                1212332.49  1232361.10  1247022.17   

                                         Y2017       Y2018  
Element                                                     
Domestic supply quantity            2088198.10  2161192.10  
Export Quantity                      182338.80   181594.80  
Fat supply quantity (g/capita/day)    10253.84    10258.69  
Feed                                 223705.68   233489.68  
Food                                1258888.28  1303841.28

In [14]:
#filtering only the Stock Variation
byElement.loc[['Stock Variation']]

Area Code  Item Code  Element Code     Y2014     Y2015  \
Element                                                                   
Stock Variation     571566   11329527      21464704  58749.83  34910.99   

                    Y2016     Y2017     Y2018  
Element                                        
Stock Variation  33140.12  54316.91  20577.91

 Y2014 had the highest with 58749.83

# Question 6:
### A pandas Dataframe with dimensions (100,3) has how many features andobservations?

#### 3 features, 100 observations 
#### features are the column names while observations are the row names

# Question 7: 
### What would be the output for?
    s = [['him','sell'], [90,28,43]]
    s[0][1][1]

In [15]:
s = [['him','sell'], [90,28,43]]
s[0][1][1]

'e'

# Question 8:
### Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’. Which of these Areas had the 7th lowest sum in 2017?

In [16]:
byArea_2017=df[['Y2017']].groupby(df['Area']).sum()
byArea_2017.head(3)

Y2017
Area              
Algeria  325644.27
Angola   229159.57
Benin    124771.22

In [17]:
# sorting values from lowest to highest
sorted_byArea=byArea_2017.sort_values('Y2017')
sorted_byArea.iloc[[6],[0]]

Y2017
Area                   
Guinea-Bissau  19102.77

Guinea Bissau is the 7th least

# Question 9:
consider the following list of tuples:
    y = [(2,4),(7,8),(1,5,9)]
    
### How would assign element 8 from the list to a variable x?

In [18]:
y = [(2,4),(7,8),(1,5,9)]


In [20]:
x = y[1][1]
x

8

# Question 10:

#### Math ia a python inbuilt module

# Question 11: 
### df.shape is used for Checking the number of rows and columns

# Question 12:
### Which of the following dataframe methods can be used to access elements acrossrows and columns?

#### df.iloc[] and df.loc[] are used to access elements acrossrows and columns

# Question 13:
### select column ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’. Which ofthese Areas had the highest sum in 2017?

In [21]:
byArea_2017=df[['Y2017']].groupby(df['Area']).sum()
byArea_2017.head(3)

Y2017
Area              
Algeria  325644.27
Angola   229159.57
Benin    124771.22

In [22]:
# sorting values from highest to lowest
sorted_byArea_dec=byArea_2017.sort_values('Y2017',ascending=False)
sorted_byArea_dec.head()

Y2017
Area                    
Nigeria       1483268.23
Egypt          866379.92
South Africa   517590.54
Ethiopia       448683.76
Morocco        388495.36

Nigeria had the highest values in 2017

# Question 14:
### Perform a groupby operation on ‘Element’. What is the total number of the sum of Processing in 2017?

In [23]:
byElement_2017=df[['Y2017']].groupby(df['Element']).sum()
byElement_2017.head()

Y2017
Element                                       
Domestic supply quantity            2088198.10
Export Quantity                      182338.80
Fat supply quantity (g/capita/day)    10253.84
Feed                                 223705.68
Food                                1258888.28

In [24]:
#filtering only the Stock Variation
byElement_2017.loc[['Processing']]

Y2017
Element             
Processing  292836.0

In 2017, the total number of the sum of Processing is was a  sum of of  292836.0 

# Question 15:
### Which year had the least correlation with ‘Element Code’?

In [27]:
df.corr(method ='pearson').sort_values('Element Code',ascending=True)

Area Code  Item Code  Element Code     Y2014     Y2015  \
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   

                 Y2016     Y2017     Y2018  
Item Code     0.020109  0.021494  0.021314  
Area Code     0.005247  0.005006  0.005665  
Y2016         1.000000  0.992785  0.992757  
Y2015         0.995739  0.988048  0.988208  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000  
Y2014         0.996081  0.995230  0.994872  
Element Code  0.023444  0.024254  0.024279

Year 2016 had a correlations of  0.023444 against Element code which is the smallest in comaprison to all the other years

# Question 16:
### What is the total Protein supply quantity in Madagascar in 2015?

In [28]:
#Madagascar production in 2015
df_Madagascar_2015 = df.loc[df['Area']=='Madagascar',['Element','Y2015']]
df_Madagascar_2015.head()

Element     Y2015
30715           Total Population - Both sexes  24234.00
30716                Domestic supply quantity      0.00
30717           Food supply (kcal/capita/day)   1922.00
30718  Protein supply quantity (g/capita/day)     43.26
30719      Fat supply quantity (g/capita/day)     22.97

In [29]:
# Madagascar protein production in 2015
madagascar_proteins=df_Madagascar_2015[df_Madagascar_2015['Element']=='Protein supply quantity (g/capita/day)']
madagascar_proteins.head()

Element  Y2015
30718  Protein supply quantity (g/capita/day)  43.26
30721  Protein supply quantity (g/capita/day)  34.04
30724  Protein supply quantity (g/capita/day)   9.23
30740  Protein supply quantity (g/capita/day)  27.26
30756  Protein supply quantity (g/capita/day)   2.14

In [30]:
# Sum of Madagascar protein production
madagascar_proteins[['Y2015']].sum(axis=0)

Y2015    173.05
dtype: float64

#### Total protein supply quantity in Madagascar in 2015 was 173.05 Protein supply quantity (g/capita/day)	

# Question 18:
### What is the total number of unique Countries in the dataset?

In [31]:
df['Area'].nunique()

49

There ae 49 unique Countries in the dataset

# Question 19:
    Given the following numpy array
    array = ([[94,89, 63],
             [93,92, 48],
             [92, 94, 56]])
### How would you select the elements in bold and italics from the array?

In [36]:
array = ([[94, 89, 63],
          [93, 92,48],
         [92, 94, 56]])

In [42]:
array[:2][1:]

[[93, 92, 48]]

# Question 20:
### What is the total sum of Wine produced in 2015 and 2018 respectively?
Hint: Perform a groupby sum aggregation on 'Item'

In [43]:
# Grouping Y2015 and Y2018  by Item using the groupby method

byItem_2015_2018=df[['Y2015','Y2018']].groupby(df['Item']).sum()
byItem_2015_2018.head()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51

In [44]:
# filtering the wine from the resulting dataframe using the loc method
byItem_2015_2018.loc[['Wine']]

Y2015    Y2018
Item                  
Wine  4251.81  4039.32

Total sum of wine produced in 2015 and 2018 is 4251.81 and 4039.32 respectively